alacritty_terminal/src/grid/storage.rs

```diff
@@ -173,25 +176,41 @@ impl<T> Storage<T> {
         }
     }
 
-    /// Truncate the invisible elements from the raw buffer
+    /// Truncate the invisible elements from the raw buffer.
     pub fn truncate(&mut self) {
         self.inner.rotate_left(self.zero);
         self.inner.truncate(self.len);
 
         self.zero = 0;
     }
 
-    /// Dynamically grow the storage buffer at runtime
+    /// Dynamically grow the storage buffer at runtime. This function tries to minimize
+    /// reallocations by using minimun step for growing internal storage. If you want to avoid this
+    /// behavior use `initialize_exact`.
     pub fn initialize(&mut self, num_rows: usize, template_row: Row<T>)
     where
         T: Clone,
     {
-        let mut new = vec![template_row; num_rows];
+        if self.len + num_rows > self.inner.len() {
+            let len = self.len;
+            self.initialize_exact(max(num_rows, MIN_INIT_SIZE), template_row);
+            // Restore proper len
+            self.len = len + num_rows;
+        } else {
+            self.len += num_rows;
+        }
+    }
 
+    /// Dynamically grow the storage buffer at runtime.
+    #[inline]
+    pub fn initialize_exact(&mut self, num_rows: usize, template_row: Row<T>)
```

Comment:
Correct me if I'm wrong, but it seems like this function is more like `always_initialize`, since it doesn't check `self.len` vs `self.inner.len` before appending. Whereas, `initialize` is more like `maybe_initialize` since it may or may not actually allocate anything. I see where `_exact` is coming from, but I don't think it makes the most sense.

Reply:
Not really, these functions are like `reserve` and `reserve_exact` from std::vec::Vec. The goal of `initialize_exact` is not to be smart, like imagine just initializing 100 elements, knowing that you wont grow any larger then that, so initializing for MIN_INIT_SIZE is pointless.